src/DefinitionResolver.php

```diff
@@ -656,13 +656,31 @@ public function resolveExpressionNodeToType($expr)
                 } else {
                     $classFqn = substr((string)$t->getFqsen(), 1);
                 }
-                $fqn = $classFqn . '->' . $expr->memberName->getText($expr->getFileContents());
-                if ($expr->parent instanceof Node\Expression\CallExpression) {
-                    $fqn .= '()';
+
+                $memberSuffix = '->' . $expr->memberName->getText($expr->getFileContents());
+                if ($expr instanceof Node\Expression\CallExpression) {
+                    $memberSuffix .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
-                    return $def->type;
+
+                // Find the right class that implements the member
+                $implementorFqns = [$classFqn];
+                while ($implementorFqn = array_shift($implementorFqns)) {
```

Comment:
Looks like this block is copied from `resolveMemberAccessExpressionNodeToFqn`, maybe there's some way the code can be reused instead of duplicated. Also the comment below about FQN now doesn't match the code.